src/Math/src/CARE.cpp

```diff
@@ -0,0 +1,191 @@
+        lhs.resize(2 * n, n);
+        rhs.resize(2 * n, n);
+    }
+};
+
+CARE::CARE()
+{
+    m_pimpl = std::make_unique<Impl>();
+}
+
+CARE::~CARE() = default;
+
+bool CARE::initialize(std::weak_ptr<ParametersHandler::IParametersHandler> handler)
+{
+    // all the parameters are optional
+    auto ptr = handler.lock();
+
+    if(ptr == nullptr)
+        return true;
+
+    ptr->getParameter("tolerance", m_pimpl->tolerance);
+    ptr->getParameter("is_verbose", m_pimpl->isVerbose);
+    ptr->getParameter("max_iterations", m_pimpl->maxIterations);
+
+    return true;
+}
+
+bool CARE::setMatrices(Eigen::Ref<const Eigen::MatrixXd> A,
+                       Eigen::Ref<const Eigen::MatrixXd> B,
+                       Eigen::Ref<const Eigen::MatrixXd> Q,
+                       Eigen::Ref<const Eigen::MatrixXd> R)
+{
+    constexpr std::string_view errorPrefix = "[CARE::setMatrices] ";
+
+    const std::size_t n = B.rows();
+    const std::size_t m = B.cols();
+
+    // check the size of the matrices
+    if (A.rows() != n || A.cols() != n)
+    {
+        std::cerr << errorPrefix << "The matrix A must be a square matrix. Expected size: " << n
+                  << " x " << n << ". Passed size: " << A.rows() << " x " << A.cols() << "."
+                  << std::endl;
+        return false;
+    }
+
+    if (Q.rows() != n || Q.cols() != n)
+    {
+        std::cerr << errorPrefix << "The matrix Q must be a square matrix. Expected size: " << n
+                  << " x " << n << ". Passed size: " << Q.rows() << " x " << Q.cols() << "."
+                  << std::endl;
+        return false;
+    }
+
+    if (R.rows() != m || R.cols() != m)
+    {
+        std::cerr << errorPrefix << "The matrix R must be a square matrix. Expected size: " << m
+                  << " x " << m << ". Passed size: " << R.rows() << " x " << R.cols() << "."
+                  << std::endl;
+        return false;
+    }
+
+    // this check may be time consuming let's consider it only when the library is compiled in debug
+    constexpr double symmetricTolerance = 1e-8;
+    assert(Q.isApprox(Q.transpose(), symmetricTolerance) && "The matrix Q must be symmetric");
+
+    m_pimpl->A = A;
+    m_pimpl->B = B;
+    m_pimpl->Q = Q;
+    m_pimpl->R = R;
+
+    m_pimpl->resizeInteralMatrices();
+
+    return true;
+}
+
+bool CARE::solve()
+{
+    // compute cholesky decomposition
+    Eigen::LLT<Eigen::MatrixXd> R_cholesky(m_pimpl->R);
+    if (R_cholesky.info() != Eigen::Success)
+    {
+        std::cerr << "[CARE::solve] The matrix R must be positive definite." << std::endl;
+        return false;
+    }
+
+    // Z represents the Hamiltonian matrix.
+    //       _                _
+    //      | A   B inv(R) B'  |
+    // Z  = |                  |
+    //      |_Q       -  A'   _|
+
+    m_pimpl->Z << m_pimpl->A, m_pimpl->B * R_cholesky.solve(m_pimpl->B.transpose()),
+                  m_pimpl->Q, -m_pimpl->A.transpose();
```

Comment:
I would suggest the use of intermediate variables to avoid the generation of temporary objects, especially for 
```
m_pimpl->B * R_cholesky.solve(m_pimpl->B.transpose()),
```
Also, for readability and for catching early mistakes, I would prefer assigning the blocks.

Reply:
It should be done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/157/commits/b3209960b6d9c6109d2ee92b2dad4638f74e5ae5

Reply:
For the  `<<` please check https://github.com/dic-iit/bipedal-locomotion-framework/pull/157/commits/40b7b4cf3d0aca65b62f67c138ec3578617758d5